dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
+            dirfd = dirfd_getter(debugger)
+            full_path = self.get_full_path(debugger, rel_file, dirfd)
+            fs_jail = fs_jail_getter(debugger)
+            return self._raw_access_check(syscall, debugger, full_path, fs_jail)
 
         return check
 
-    def _file_access_check(
-        self, rel_file, debugger, fs_jail, dirfd=AT_FDCWD
-    ) -> Tuple[str, Optional[ErrnoHandlerCallback]]:
-        # Either process called open(NULL, ...), or we failed to read the path
-        # in cptbox.  Either way this call should not be allowed; if the path
-        # was indeed NULL we can end the request before it gets to the kernel
-        # without any downside, and if it was *not* NULL and we failed to read
-        # it, then we should *definitely* stop the call here.
-        if rel_file is None:
-            return '(nil)', ACCESS_EFAULT
+    # False denotes normal behaviour, but a disallowed file
+    def _raw_access_check(self, syscall: str, debugger: Debugger, file: str, fs_jail: FilesystemPolicy) -> bool:
+        # file can be non-normalized, which is required to make sure we are checking the right file
 
-        try:
-            file = self.get_full_path(debugger, rel_file, dirfd)
-        except UnicodeDecodeError:
-            log.exception('Unicode decoding error while opening relative to %d: %r', dirfd, rel_file)
-            return '(undecodable)', ACCESS_EINVAL
-
-        # We want to ensure that if there are symlinks, the user must be able to access both the symlink and
-        # its destination. However, we are doing path-based checks, which means we have to check these as
-        # as normalized paths. normpath can normalize a path, but also changes the meaning of paths in presence of
-        # symlinked directories etc. Therefore, we compare both realpath and normpath and ensure that they refer to
-        # the same file, and check the accessibility of both.
-        #
-        # This works, except when the child process uses /proc/self, which refers to something else in this process.
-        # Therefore, we "project" it by changing it to /proc/[tid] for computing the realpath and doing the samefile
-        # check. However, we still keep it as /proc/self when checking access rules.
-        projected = normalized = '/' + os.path.normpath(file).lstrip('/')
-        if normalized.startswith('/proc/self'):
-            file = os.path.join(f'/proc/{debugger.tid}', os.path.relpath(file, '/proc/self'))
-            projected = '/' + os.path.normpath(file).lstrip('/')
-        elif normalized.startswith(f'/proc/{debugger.tid}/'):
-            # If the child process uses /proc/getpid()/foo, set the normalized path to be /proc/self/foo.
-            # Access rules can more easily check /proc/self.
-            normalized = os.path.join('/proc/self', os.path.relpath(file, f'/proc/{debugger.tid}'))
-        real = os.path.realpath(file)
+        # We need to project /proc/self.
```

Comment:
It's worth keeping the original comment here

Reply:
I'd appreciate leaving this code as it was, it's harder to read now that one has to keep going to read short functions (that aren't reused elsewhere for it to be worth it)

Reply:
(it's not obvious from the function names the nuance of what they actually do, and the nuance is complicated enough that encoding it in a name seems impossible)